internal/handlers/common/saslstart_plain.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// saslStartPlain extracts username and password from PLAIN `saslStart` payload.
+func saslStartPlain(doc *types.Document) (string, string, error) {
+	var payload []byte
+
+	// some drivers send payload as a string
+	stringPayload, err := GetRequiredParam[string](doc, "payload")
+	if err == nil {
+		if payload, err = base64.StdEncoding.DecodeString(stringPayload); err != nil {
+			return "", "", lazyerrors.Error(err)
+		}
+	}
+
+	// most drivers follow spec and send payload as a binary
+	binaryPayload, err := GetRequiredParam[types.Binary](doc, "payload")
+	if err == nil {
+		payload = binaryPayload.B
+	}
+
+	if payload == nil {
```

Comment:
This `if` looks unnecessary to me. We either have no error and some payload, or an error and `nil` payload. Or am I missing something?

Reply:
Yes that's a bit confusing. @AlekSi why can't we check the error here?

Reply:
I think you ask two different questions, though.

`if payload == nil` could be replaced with `if err != nil`.

However, one of those two `if`s is necessary there

Reply:
Then maybe it's better to replace the whole constructions of ifs with switch-case to increase readability?

Reply:
For readability, switch statement might be superior but what we have is fine too :hugs:

Reply:
@chilagrow I tried a few various approaches, and in the end, the current one looked better to me, with less cyclomatic complexity etc. But if you see a clearer approach, could you please suggest a particular code?

Reply:
Yeah I agree about a bit more complexity, in exchange for more readability. Happy to keep what it is for now. 

```
	var payload []byte
	var err error

	v, _ := doc.Get("payload")
	if v == nil {
		return "", "", commonerrors.NewCommandErrorMsgWithArgument(
			commonerrors.ErrBadValue,
			"required parameter \"payload\" is missing",
			"payload",
		)
	}

	switch v := v.(type) {
	case string:
		// some drivers send payload as a string
		if payload, err = base64.StdEncoding.DecodeString(v); err != nil {
			return "", "", commonerrors.NewCommandErrorMsgWithArgument(
				commonerrors.ErrBadValue,
				fmt.Sprintf("Invalid payload: %v", err),
				"payload",
			)
		}
	case types.Binary:
		// most drivers follow spec and send payload as a binary
		payload = v.B
	default:
		return "", "", commonerrors.NewCommandErrorMsgWithArgument(
			commonerrors.ErrBadValue,
			fmt.Sprintf("required parameter \"payload\" has type %T (expected types.Binary)", v),
			"payload",
		)
	}
```